crates/bevy_core_pipeline/src/bloom/bloom.wgsl

```diff
@@ -1,138 +1,137 @@
+// Bloom works by creating an intermediate texture with a bunch of mip levels, each half the size of the previous.
+// You then downsample each mip (starting with the original texture) to the lower resolution mip under it, going in order.
+// You then upsample each mip (starting from the smallest mip) and blend with the higher resolution mip above it (ending on the original texture).
+//
+// References:
+// * http://www.iryoku.com/next-generation-post-processing-in-call-of-duty-advanced-warfare
+// * https://learnopengl.com/Guest-Articles/2022/Phys.-Based-Bloom
+
 #import bevy_core_pipeline::fullscreen_vertex_shader
+#import bevy_core_pipeline::tonemapping
 
-struct BloomUniforms {
-    threshold: f32,
-    knee: f32,
-    scale: f32,
-    intensity: f32,
+struct BloomDownsamplingUniforms {
+    threshold_precomputations: vec4<f32>,
     viewport: vec4<f32>,
 };
 
 @group(0) @binding(0)
-var original: texture_2d<f32>;
+var input_texture: texture_2d<f32>;
 @group(0) @binding(1)
-var original_sampler: sampler;
-@group(0) @binding(2)
-var<uniform> uniforms: BloomUniforms;
-@group(0) @binding(3)
-var up: texture_2d<f32>;
+var s: sampler;
 
-fn quadratic_threshold(color: vec4<f32>, threshold: f32, curve: vec3<f32>) -> vec4<f32> {
-    let br = max(max(color.r, color.g), color.b);
-
-    var rq: f32 = clamp(br - curve.x, 0.0, curve.y);
-    rq = curve.z * rq * rq;
+#ifdef FIRST_DOWNSAMPLE
+@group(0) @binding(2)
+var<uniform> uniforms: BloomDownsamplingUniforms;
+
+// https://catlikecoding.com/unity/tutorials/advanced-rendering/bloom/#3.4
+fn soft_threshold(color: vec3<f32>) -> vec3<f32> {
+    let brightness = max(color.r, max(color.g, color.b));
+    var softness = brightness - uniforms.threshold_precomputations.y;
+    softness = clamp(softness, 0.0, uniforms.threshold_precomputations.z);
+    softness = softness * softness * uniforms.threshold_precomputations.w;
+    var contribution = max(brightness - uniforms.threshold_precomputations.x, softness);
+    contribution /= max(brightness, 0.00001); // Prevent division by 0
+    return color * contribution;
+}
+#endif
 
-    return color * max(rq, br - threshold) / max(br, 0.0001);
+// http://graphicrants.blogspot.com/2013/12/tone-mapping.html
+fn karis_average(color: vec3<f32>) -> f32 {
+    let luma = tonemapping_luminance(rgb_to_srgb_simple(color)) / 4.0;
```

Comment:
I'm not sure either.

Reply:
The implementation is faithful to the source material. Input signal is in linearized sRGB and luma is a weighted sum or sRGB components. We convert our image to sRGB and perform the luminance function (which is the same as luma but with gamma-corrected input). I don't know why this works considering we are trying to process physical light and avoid any human-based measurements like luminance and, even more so, luma, but this function seems to do the trick somehow. I am pretty positive this will lead to artifacts if the input color space changes (for example to ACES 2065-1) but at the moment it works. I think the approach the original developers took here was somewhat similar to the one described in [this video](https://youtu.be/wt2yYnBRD3U?t=947) (15:47 - 16:15)

Reply:
:)

Reply:
For clarity, and because I can't resolve these discussions, I consider this discussion resolved well enough for now.